### ISSUE TITLE
Column-Level Lineage

### ISSUE THREAD

**Issue by reata** [reata/sqllineage#103]
Since @ekimd has started the effort for column-level lineage analysis. I create this ticket to track all the questions we have to answer before we dive into implementation details. Considering our pure code analysis approach without involving metadata, I believe we have several design choices to make.

Question No.1: What's the data structure to represent Column-Level Lineage.
Currently we're using `DiGraph` in library `networkx` to represent Table-Level Lineage, with table as vertex and table-level lineage as edge, which is pretty straight forward. After changing to Column-Level, what's the plan?

Question No.2: How do we deal with `select *`
```
INSERT OVERWRITE tab1
SELECT * FROM tab2;
```
In this case, we don't know which columns are in tab2.

Question No.3: How do we deal with column without table/alias prefix in case of join.
```
INSERT OVERWRITE tab1
SELECT col2
FROM tab2
JOIN tab3
ON tab2.col1 = tab3.col1
```
In this case, we don't know whether col2 is coming from tab2 or tab3.

Question No.4: How do we visualize column-level lineage?

**Comment by reata** [comment]
@ekimd Would you share your design choices here?

Just had a tough week. But I think I shall be able to work with you on this feature from now on. This feature is huge, but I'm also highly looking forward to it.

**Comment by coldmoutain** [comment]
hello，reata,I may give  some suggestions  on Q4, when I visualize the column-level lineage ,I  used Neo4j to visualize It. so may we use the Knowledge Graph like  method to visualize it

**Comment by coldmoutain** [comment]
As for Q2&Q3，Pure code ,actually,Text  analysis could not solve Q2 and Q3. Information we get from pure code are limited ,so we need more information from metadata.  But most users don`t have access to metadata database . in my humble opinion,if users wannna get Column-Level Lineage ,at least,they need connect to the databases where their codes run. we can use the keyword "describe" to get the source table structure.it`s a simple way

**Comment by reata** [comment]
Two key design principles as I'm considering this feature:

- **sqllineage will stay primarily as a static code analysis tool**, so we must tolerate information missing when doing column-level lineage. Maybe somewhere in the future, we can provide some kind of plugin mechanism to register metadata as a supplement to refine the lineage result, but in no way will we depend solely on metadata. See  [docs](https://sqllineage.readthedocs.io/en/latest/behind_the_scene/dos_and_donts.html):
- **I don't want column-level lineage DAG to be a separate graph from table-level DAG**. There should be one unified DAG. Either  1) we build a DAG to the granularity of column so with some kind of transformation, we can derive table-level DAG from it. To use an analogy of relational data, it's like building a detail table, with the ability to aggregate up to a summary table. Or 2) we build a property graph (see JanusGraph [docs](https://docs.janusgraph.org/basics/schema/) for reference), with Table and Column as two kinds of Vertex, and two type of edges, one for column to table relationship mapping, one for column-level as well as table-level lineage. 


So here's a possible approach each answer:
Q1: See design principle No. 2 for two possible data structure. I would prefer property graph for this. But I'm not sure if `networkx` can support it.
Q2: Add an virtual column `*` for each Table as a special column. So the lineage is tab2.* -> tab1.*
Q3: Add two edges, tab2.col2 -> tab1.col2, tab3.col2 -> tab1.col2. Meanwhile, these two edges should be marked so that later in visualization, they can be drawn differently, like in dot line.
Q4: I would very much like to visualize column-level lineage like [GCP Cloud Data Fusion](https://cloud.google.com/data-fusion/docs/tutorials/lineage). Staying with Matplotlib, or turning to a JavaScript-based approach, I'm both happy. Also as you guys suggest, we can add a io module to export our graph to GML so that people can choose to visualize it in graph database like neo4j for their own.

**Comment by ekimd** [comment]
Sorry for the slow reply, been really busy.

This is why I wasn't sure if this merge request is appropriate because it
doesn't 100% fit the existing design model.  I agree with all of your
comments below.  In my specific circumstance, I was looking to essentially
trace column lineage as it moved from tabled to table through many
transforms, and there were no "select *" statements in any of the SQL.  GCP
cloud data fusion view would be really nice, but I'm not sure how hard it
would be to implement.  That's why I like the idea of searializing the
graph and then letting people choose their viewing tool (I used Cytoscape
and it worked decently).

**Comment by reata** [comment]
> Sorry for the slow reply, been really busy. This is why I wasn't sure if this merge request is appropriate because it doesn't 100% fit the existing design model. I agree with all of your comments below. In my specific circumstance, I was looking to essentially trace column lineage as it moved from tabled to table through many transforms, and there were no "select *" statements in any of the SQL. GCP cloud data fusion view would be really nice, but I'm not sure how hard it would be to implement. That's why I like the idea of searializing the graph and then letting people choose their viewing tool (I used Cytoscape and it worked decently).

Haha, fully understood, always full of paperwork towards the end of the year, at least that's my situation right now. I think I'll be able to spare some time on this ticket during Christmas and New Year.

**Comment by wowpoppy** [comment]
Q4:
Don't have much to say about visualizing the graph, however, it would be great if `SQLLineage` was not only a `CLI`-first tool with graphic visualization abilities but also a powerful library with a decent and well-documented programmatic `API`.

Q2:
As of dealing with `SELECT *`, wouldn't it be okay to report that no data on column-level lineage is available? Seems like too much for a `SQL` parsing library that does not actually connect to the database and therefore cannot retrieve metadata to get information about columns.

Q3:
As mentioned above, any of provided source tables might be considered as __possible__ sources.

I'm currently working on a project that requires column-level lineage so I might fork @ekimd's work to fulfill my project's requirements.

**Comment by reata** [comment]
Hi @wowpoppy ,
@ekimd  approach is actually very enlightening to me. So if you feel the same, please do refer to his work.

For SQLLineage to become more than a CLI-first tool. Right now I'm doing some front end visualization work (with react/redux/cytoscape/monaco editor, to name a few JS package used here). The intention behind this is: 1) to get rid of graphviz so that it will be more friendly to windows user; 2) lay the foundation for future column-level visualization.

It's taking me more time than I first estimated. Hopefully I'll get front end coding part done this month and come back to column-level lineage as well as programmatic API soon after.

You're right. Currently I'm intentional hiding some of the Python API under the hood. On the contrary, I'm assuming the CLI interface is mature enough so this is the part getting more document.

**Comment by reata** [comment]
With 6308b50e0b087e1bdab722dd531282a169131f4b splitting the logic into different handlers, we now have SourceHandler, TargetHandler and CTEHandler to handle table level lineage. They're subclass of NextTokenBaseHandler, an abstract class to address an extract pattern when a specified token indicates we should extract something from next token. 

After this refactor, I'm considering introduce a new abstract class SectionTokenBaseHandler to address an extract pattern when we should extract something from a section of tokens, with ColumnHandler being its only subclass for now. For most of the case, anything between keywords SELECT and FROM should be our target columns. (This is also how @ekimd attacks this problem.) Let's see if this approach works.

My answer on 2020.12 still holds. I will use a property graph, use virtual column *, tolerate vague column reference. And with current JavaScript-based visualization, we should be able to visualize column level lineage without much effort.

Development is on develop branch, and to my estimate we shall need following stages to implement column level lineage:

1. Atomic column logic handling: alias, case when, function, expression
2. Column to table assignment in case of table join
3. Subquery
4. Assemble Statement Level lineage into multiple statements DAG.

Considering the complexity of this feature, target release date is early October 2021.

As of now, stage 1 is imperfectly done with 25116cb62c2b20d0c37fcd0377871f9272c68d94 . Of course we shall come back to this with various edge cases.

**Comment by dingtianlu** [comment]
Sorry, Hello, can Column-Level Lineage be released as scheduled？I am looking forward to this feature。

**Comment by reata** [comment]
> Sorry, Hello, can Column-Level Lineage be released as scheduled？I am looking forward to this feature。

Yes, everything is on track.

**Comment by wangqinghuan** [comment]
Do we support analyze "CREATE TABLE" statement? For Spark sql and Flink sql, we usually use "CREATE TABLE" statement to connect external table firstly, then use "SELECT" or "INSERT" statement to operate table. In these cases which contains "CREATE TABLE" statements, we can analyze column lineage for Q2 and Q3 because we can derive original field from "CREATE TABLE" statements. And  Flink sql and Spark sql is very common for data etl and process in BigData, it's very helpful if we support Flink sql and spark sql analyze.

**Comment by reata** [comment]
> Do we support analyze "CREATE TABLE" statement? For Spark sql and Flink sql, we usually use "CREATE TABLE" statement to connect external table firstly, then use "SELECT" or "INSERT" statement to operate table. In these cases which contains "CREATE TABLE" statements, we can analyze column lineage for Q2 and Q3 because we can derive original field from "CREATE TABLE" statements. And Flink sql and Spark sql is very common for data etl and process in BigData, it's very helpful if we support Flink sql and spark sql analyze. @reata

If you check the current functionality for table level lineage, you'll find "CREATE TABLE" statement is already supported.
```bash
$ sqllineage -e "CREATE TABLE db.tab1 AS SELECT * FROM db.tab2"
Statements(#): 1
Source Tables:
    db.tab2
Target Tables:
    db.tab1
```
Going to column level, I don't see any potential issue that will lead us to drop support for "CREATE TABLE" statement.

**Comment by reata** [comment]
With 9bb23cf996d7a05ba7cf8ea516fc9a5a8abf26b7 merged into master, we now have a method for subquery recognition. That will help here to identify whether the column selected is within a subquery or at the top level.

**Comment by reata** [comment]
Having commit c8682b4128447d03c2b1ba84c3714fbba9f57882, it's now possible to extract column level lineage with subquery involved. Next step is for column to table assignment in case of table join

**Comment by reata** [comment]
Column to table assignment in case of table join is done with commit ecd70ca925efbd7ec08e8f7190e585855bcda13c.
An implementation correction here: SectionTokenBaseHandler approach is abandoned. Column handler is still based on NextTokenBaseHandler (column token followed by keyword SELECT) plus a end-of-(sub)query hook. Because only until end of query could we know all the source tables. If we don't have all the source tables and their alias, we can't assign the column to table correctly.

**Comment by reata** [comment]
e7e4cb0a872b18b3f168567390bf9e94bd56da53 is an important refactor which introduces the abstraction of LineageHolder, we now have three lineage holders each at different level:

- SubQueryLineageHolder holds lineage at subquery (and DML query) level. For the moment it's almost the same as StatementLineageHolder except that we don't have table drop or rename for subquery.
- StatementLineageHolder holds lineage at statement level. It's a replacement for the old LineageResult
- SQLLineageHolder holds lineage at SQL files (or SQL commands with multiple statements) level. It's a replacement for the old CombinedLineageResult

combiners module contains combine functions to bridge the LineageHolder at different level.

With the new abstraction, next I'll delicately design the inner data structure of each holder. Right now it's still two graphs for column and table respectively thus redundancy exists. Hopefully I'll be able to combine the two graphs into one property graph as designed and still capable of exposing the current interface to end user.

**Comment by reata** [comment]
5a6f9397fb053a03274b59d0598e7efb93d4dc9a unified all holders to use a property graph as the sole internal data structure. And to make the graph reading and writing more inherent, combiners module is removed, replaced by a "of" staticmethod in holder class.

A command line option "-l column" is added to expose the column lineage. Now that all the implementation is done based on the design, next I'll spend several days doing all kinds of tests as well as making the command interface more user-friendly. (Now it's only source column -> target column pairs with intermediate column hidden, maybe a column path is the preferred way.)

I'll create another ticket to trace JavaScript visualization for column-level lineage and limit the scope of this ticket to command line interface only.

**Comment by reata** [comment]
Correction: adding JavaScript visualization for column-level lineage back to the scope of this ticket.

Using cytoscapejs's [compound nodes](https://js.cytoscape.org/#notation/compound-nodes) functionality, visualize column level lineage is easier than I expected. Now with #170 merged, visualization is ready for test.

For the same sql provided above, this is the visualization. (we offer a table level/column level switch in lineage view, what's shown below is column level only but table level is still available)

![null](https://user-images.githubusercontent.com/8022906/136662773-25ba9f5b-8bfa-412c-b391-c4a284d92d29.jpg)

Key conclusion:

- Column Level Lineage feature with visualization is now ready for test. Just install the code from master branch with `pip install git+https://github.com/reata/sqllineage.git
`
- Release date still set at next month, since we need time to do more test as well as refine document. But consider all the heavy work as done.

**Comment by reata** [comment]
> Pretty sweet,
> 
> I notice when examining column level lineage, We trace the columns from source tables into CTEs for example but we do not trace them to the final query beneath the CTEs which is a very important piece. Is this perhaps a regression?
> 
> ![image](https://user-images.githubusercontent.com/41213451/136902714-ab582225-5239-4cee-ab53-bcf190c4b241.png)
> 
> ![image](https://user-images.githubusercontent.com/41213451/136903020-f032153b-cf58-4944-bc57-a3678f6610fd.png)

This is expected behavior. Table-wise, here we don't have a target table, only two source table and two CTE. If you change the SELECT statement to a INSERT OVERWRITE table xxx SELECT ..., then the lineage graph will be more sensible.

The only change I made to CTE is to model them as SubQuery instead of Table. SubQuery only shows up in column level lineage graph, not for table level. Let me know if that's not the case.

**Comment by reata** [comment]
Known Issues Update:

- Exception for complex CASE clause. Token after both WHEN and THEN can be nested.
- Window Function with parameter. Now row_number() OVER ... and rank() OVER ... are handled. But window function like max(col) OVER ..., sum(col1) OVER ... will get wrong result.

@eeroel  is working on #172 to get these fixed.

**Comment by reata** [comment]
Closing this issue as we're ready for v1.3.0 release.

Split the remaining cases to two separate issues and targeting get them fixed in v1.3.1:

- #183 
- #184 

Thanks everyone for helping with this feature.
